### ISSUE TITLE
haxelib reads haxe files in current directory

### ISSUE THREAD

**Issue by hughsando** [HaxeFoundation/haxelib#159]
When haxelib uses  "compile and run" to run, the haxe file in the current directory take preference over the haxe files from haxe_std_path.  Particularly, haxelib can't be from from the nme directory because haxe.Timer from nme gets pulled in and then causes an error.
To test, in a fresh directory, create a file "haxe/Timer.hx" with some junk in it, and try to run haxelib.

**Comment by back2dos** [comment]
I think the best we can do is to change the cwd to where the haxelib script is and then --run it there. That should avoid such collisions. Could you look into that @jasononeil ?

**Comment by jasononeil** [comment]
@back2dos It turns out the Haxe Makefile was already doing something similar (change directory before running), so I made the `/usr/bin/haxelib` shell script generated by selfupdate do the same.

Can you check if this will affect the Windows update mechanism, with "update.hxml" etc?

**Comment by jasononeil** [comment]
Alright, the "change directory first" solution is problematic.  Because we change the cwd, all of the haxelib commands which run on the current cwd are broken.  (Everything from "install build.hxml" to "submit" to "newrepo").  

I see a few options:

1. Keep changing the directory first, but use some environment variable "HAXELIB_CWD" or similar.  To do this we'd need to:
    * Modify Main.hx to use the environment variable instead of the `Sys.getCwd()` (if it exists)
    * Modify `updateSelf` to add the environment variable into the shell script
    * Modify Haxe's Makefile as well, (which generates it's own shell script)
    * Check if other methods for installing Haxe are affected (PPA, homebrew, OpenFL installer, hvm)
2. Add a flag to the Haxe compiler that specifies the CWD should not be used as a class path.
3. Use a pre-compiled version of haxelib rather than use "--run", similar to Windows.

I'm not familiar with the compiler, but I feel like (2) would be the least work and the least risk of breaking things for users.   I can do (1) if that is the desired approach.  (3) I guess would look similar to what we do for Windows currently.  It would mean we re-add the dependency on neko to have a functioning Haxe installation.

For now I'm going to revert the commit I made above.

**Comment by hughsando** [comment]
I think 2 could be useful elsewhere maybe, so a good option.
I like 3 because it solves the bootstrap problem if you are developing
haxelib itself.

Another option might be to explicitly add $HAXE_STD_PATH as a class path -
would this then take priority over "." rather than the other way around?

I would certainly go for option 2 over option 1.

**Comment by jasononeil** [comment]
I agree that option 1 is the one to be avoided.

So that leaves:

* Option 2: Haxe flag to not use "." as classpath
* Option 3: Pre-compiled haxelibs on Unix platforms
* Option 4: Some way of setting a classpath (or HAXE_STD_PATH) as higher priority than "."

@ncannasse do you want to decide?

**Comment by back2dos** [comment]
Wasn't neko on linux always a tricky subject, or is that a thing of the past?

Anyway, I think there's a solution where the haxelib script does something like this:

```
var previousCwd = getCwd();
setCwd(directoryOfThisScript());
command('haxe', ['-cp', 'haxelibSource', '--run', 'tools.haxelib.Main', '-cwd', previousCwd].concat(Sys.args());
setCwd(previousCwd);
```

We'd just have to support a -cwd argument for haxelib (that'll just setCwd()) and write the above in bash or whatever and things should work, right?

**Comment by hughsando** [comment]
Might not work with a local .haxelib directory, but then you may be asking
for trouble if you reimplement haxelib in a .haxelib directory. But maybe
this is not so crazy.

But I think this should work:

command('haxe', ['-cp', 'haxelibSource', '-cp',
Sys.getEnv("HAXE_STD_PATH"), '--run',
'tools.haxelib.Main'].concat(Sys.args());

I still like the --classpath-no-cwd option.  Cpp uses #inclucde <a.h>
vs #include "a.h" to differentiate whether . is in the include-path,
so I think it is a useful option.

**Comment by back2dos** [comment]
Ok, @jasononeil, the `-cwd` is in place. How are your bash skills? ;)

**Comment by jasononeil** [comment]
This is done as far as haxelib is concerned.  (Still have to update the Makefile on the Haxe installer, but probably not until they're ready to do another sync of the Haxelib repo...) 

But I want to explore option 3 further - using binaries on unix platforms, rather than "--run".  

One issue I ran into was `sys.io.Process( 'which', ['haxelib'] )` not working anymore on 3.1.3, but working correctly on 3.2.  But it used to work on 3.1.3, so now I'm just confused.  There's also the no-macros thing, and the fact that the Ubuntu PPA, the binary downloads, and the OpenFL installer all use precompiled binaries.

Should I open another issue to start that discussion?

**Comment by back2dos** [comment]
Yep, another issue would be good. We could start there by documenting the differences between --run --interp and precompiled neko. Can you close this issue if it's dealt with?

**Comment by hughsando** [comment]
Maybe you fixed this post 3.2 rc2 release?  The following fixed worked for me on mac - adding HAXE_STD_PATH explicitly in the shell script:
#!/bin/sh
exec haxe -cp "/usr/lib/haxe/lib/haxelib_client/3,2,0-rc,2/src/" -cp "$HAXE_STD_PATH" -D haxelib_client=3.2.0-rc.2 --run tools.haxelib.Main "$@"

Hugh

**Comment by ibilon** [comment]
There's no run script anymore, so that should be fixed (if I understand the issue correctly).

**Comment by nadako** [comment]
Indeed!
